subsys: debug: Added full CPU utilization monitor

koffes opened the pull request:
- Added a module and tests to detect full CPU load

koffes commented:
This PR resolves #71945. We may need to discuss a bit when it comes to multicore architectures.

nashif commented:
you need to rebase

nordic-krch commented:
I wonder if it would not be lighter and simpler to reuse already existing periodic cpu load logging functionality. Move it to k_timer instead of k_work and add option to report load only if load is above given configurable threshold. When threshold is set then it would report it as LOG_WRN (or printk if logging is disabled). Using current approach requires enabling thread stats which increases code size and that may be an issue for smaller cores (e.g. flpr).

koffes commented:
PR has been rebased.
Thanks for the input @nordic-krch. I would like more people to weigh in on your suggestion on "periodic cpu load". I can rewrite to use that, but would like to know if it is required. There will be changes there as well as you say, changing to k_timer will influence all other applications. I'd prefer to minimize impact on existing features. Using LOG_WRN is not an option, as that will be scheduled out.

koffes committed:
debug: Added CPU usage callback

- Added a CPU load callback with threshold
- Changed cpu_load to use k_timer instead of k_work

Signed-off-by: Kristoffer Rist Skøien <kristoffer.skoien@nordicsemi.no>

koffes commented:
@nordic-krch : re-written based on your comment.

nordic-krch commented on tests/subsys/debug/cpu_load/src/main.c:
```diff
@@ -94,6 +94,49 @@ ZTEST(cpu_load, test_periodic_report)
+void high_load_cb(uint8_t percent)
+{
+	last_cpu_load_percent = percent;
+	TC_PRINT("percendt %u", percent);
```
```suggestion
	TC_PRINT("percent %u", percent);
```

koffes replied:
For debugging. Will be removed.

nordic-krch commented on include/zephyr/debug/cpu_load.h:
```diff
@@ -48,6 +49,23 @@ int cpu_load_get(bool reset);
+ *
+ * @retval 0 - Callback registered/cancelled.
+ * @retval -EINVAL if the threshold is invalid.
+ *
```
redundant line

nordic-krch commented on include/zephyr/debug/cpu_load.h:
```diff
@@ -48,6 +49,23 @@ int cpu_load_get(bool reset);
+/** @brief Optional registration of callback when load is greater or equal to the threshold.
+ *
+ * Submitting NULL pointer will cancel the callback.
+ *
```
can you add parameter description?

nordic-krch commented on include/zephyr/debug/cpu_load.h:
```diff
@@ -48,6 +49,23 @@ int cpu_load_get(bool reset);
 
+/* Optional callback for cpu_load_cb_reg
+ *
+ * This will be called from the k_timer expiry_fn.
```
```suggestion
 * This will be called from the k_timer expiry_fn used for periodic logging. CONFIG_CPU_LOAD_LOG_PERIODICALLY must be configured to a positive value.
```

nordic-krch reviewed (commented):
Looks ok. I've opened #88222 so it will be possible to enable periodic logging but set cpu_load log level below INF and use periodic callback only for tracking if threshold is exceeded.

koffes commented:
This has been idle for some time. I will rebase again, but @nashif can you have a look?

dkalowsk commented:
@dcpleung can you give a look at this as well?  @andyross not sure if you'd want to review here as well
